### PR TITLE
[FrameworkBundle] add attribute stamps

### PR DESCRIPTION
**Pull request by walidboughdiri**
| Q             | A
| ------------- | ---
| Branch?       | 4.4 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | n/a   
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by Tobion** [comment]
It's a new feature that cannot go in 4.3

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 1]
```diff
@@ -397,18 +398,18 @@ protected function isCsrfTokenValid(string $id, ?string $token): bool
...
      * Dispatches a message to the bus.
      *
-     * @param object|Envelope $message The message or the message pre-wrapped in an envelope
-     *
+     * @param object|Envelope  $message The message or the message pre-wrapped in an envelope
+     * @param StampInterface[] $stamps
      * @final
      */
-    protected function dispatchMessage($message): Envelope
+    protected function dispatchMessage($message, array $stamps = []): Envelope
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: this should be done in a BC way: the argument should be commented out and accessed using func_get_arg - see the code base for similar things.

Comment (walidboughdiri) [reply src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: @nicolas-grekas thks for comment, changes made for BC way

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 2]
```diff
@@ -397,18 +398,22 @@ protected function isCsrfTokenValid(string $id, ?string $token): bool
     /**
      * Dispatches a message to the bus.
      *
-     * @param object|Envelope $message The message or the message pre-wrapped in an envelope
-     *
+     * @param object|Envelope  $message The message or the message pre-wrapped in an envelope
+     * @param StampInterface[] $stamps
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: please keep a blank line after this one

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 3]
```diff
@@ -423,13 +428,11 @@ protected function addLink(Request $request, Link $link)
         if (!class_exists(AddLinkHeaderListener::class)) {
             throw new \LogicException('You can not use the "addLink" method if the WebLink component is not available. Try running "composer require symfony/web-link".');
         }
-
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: should be kept

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 4]
```diff
@@ -423,13 +428,11 @@ protected function addLink(Request $request, Link $link)
...
         if (null === $linkProvider = $request->attributes->get('_links')) {
             $request->attributes->set('_links', new GenericLinkProvider([$link]));
 
             return;
         }
-
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: should be kept

**Review by nicolas-grekas — APPROVED** [review]
LGTM, just minor CS comments to fix before merging.

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php (hunk 2 above)
Comment (walidboughdiri) [reply src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: done

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 5]
```diff
@@ -397,18 +398,23 @@ protected function isCsrfTokenValid(string $id, ?string $token): bool
...
      * Dispatches a message to the bus.
      *
-     * @param object|Envelope $message The message or the message pre-wrapped in an envelope
+     * @param object|Envelope  $message The message or the message pre-wrapped in an envelope
+     * @param StampInterface[] $stamps
      *
      * @final
      */
-    protected function dispatchMessage($message): Envelope
+    protected function dispatchMessage($message/*, array $stamps = []*/): Envelope
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: As the method is marked as being `@final`, we don't need the BC break code. We can add the new argument directly.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php (hunk 5 above)
Comment (walidboughdiri) [reply src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: @fabpot BC break code removed.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @walidboughdiri.

**Pull request merged**